pyccel/ast/core.py

```diff
@@ -1313,6 +1329,118 @@ def declarations(self):
         return [Declare(i.dtype, i) for i in self.variables]
 
 
+#==============================================================================
+class Iterable(Basic):
+    """
+    Wrapper around iterable types helping to convert between those
+    types and a range (necessary in low level languages, e.g. C and Fortran)
+
+    Paramaters
+    ----------
+    iterable : acceptable_iterator_type
+                The iterator being wrapped
+    """
+    indexless_types = (PythonRange, PythonEnumerate)
+    acceptable_iterator_types = (Variable, PythonMap, PythonZip, PythonEnumerate, Product, PythonRange)
+    __slots__ = ('_iterable','_indices','_num_indices_required')
+    _attribute_nodes = ('_iterable','_indices')
+
+    def __init__(self, iterable):
+        self._iterable = iterable
+        self._indices  = None
+
+        if isinstance(iterable, self.indexless_types):
+            self._num_indices_required =  0
+        elif isinstance(iterable, Product):
+            self._num_indices_required = len(iterable.elements)
+        elif isinstance(iterable, self.acceptable_iterator_types):
+            self._num_indices_required = 1
+        else:
+            raise TypeError("Unknown iterator type")
+
+        super().__init__()
+
+    @property
+    def num_indices_required(self):
+        """ Number of indices which must be generated in order to
+        convert this iteratable to a range
+        """
+        return self._num_indices_required
+
+    def set_indices(self, *indices):
+        """ Set the indices used to define the iterator/target from the range
+        """
+        self._indices = indices
+
+    def get_assigns(self, target):
+        """ Returns a list containing any assigns necessary to initialise
+        the loop iterators/targets when using a range iterable
+
+        Parameters
+        ----------
+        target : Variable or list of Variables
+                 The index(es) over which the loop iterates
+
+        Results
+        -------
+        assigns : list of Assign
+                  The assignments necessary to define target
+        """
+        iterable = self._iterable
+        if isinstance(iterable, PythonRange):
+            return []
+        ranges = self.get_target_from_range()
```

Comment:
Probably yes, but in that case I don't know how to differentiate it from the argument `target`. Maybe `range_element` ?
If we are looking at:
```
for ai in a:
```
Then `ranges` (or `range_element`) is `a[i]` and `target` is `ai`

Reply:
Better? 23c19c9

Reply:
I think these will all work: `elements`, `indexed`, `values`...  Maybe the function also needs a different name

Reply:
So `range_element` is ok? Maybe, I don't think iterable/iterator/target is particularly clear. Do you have a suggestion?

Reply:
I think `iterable` is totally fine, as it is a Python concept. I am always confused by `target` as it could mean anything. I'm going to check the Python documentation

Reply:
I think `target` is the place where the result is saved. In `For` it is also called `iterator` but I find the use of `iterator` and `iterable` to be quite confusing as the names are so close

Reply:
I agree that `iterator` is definitely not a good name for the left-hand side of `ai = a[i]`

Reply:
Well, I cannot find a set of names which convince me completely... 😅 
Maybe add a few comments to the code for now?